pkg/util/certificates/csr.go

```diff
@@ -38,3 +47,96 @@ func ParseCertificateRequestObject(obj *certificates.CertificateSigningRequest)
 	}
 	return csr, nil
 }
+
+// NewCertificateRequest generates a PEM-encoded CSR using the supplied private
+// key file, subject, and SANs. If the private key file does not exist, it generates a
+// new ECDSA P256 key to use and writes it to the keyFile path.
+func NewCertificateRequest(keyFile string, subject *pkix.Name, dnsSANs []string, ipSANs []net.IP) ([]byte, error) {
+	var privateKey interface{}
+
+	if _, err := os.Stat(keyFile); os.IsNotExist(err) {
+		privateKey, err = ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
+		if err != nil {
+			return nil, err
+		}
+
+		ecdsaKey := privateKey.(*ecdsa.PrivateKey)
+		derBytes, err := x509.MarshalECPrivateKey(ecdsaKey)
+		if err != nil {
+			return nil, err
+		}
+
+		pemBlock := &pem.Block{
+			Type:  "EC PRIVATE KEY",
+			Bytes: derBytes,
+		}
+
+		err = ioutil.WriteFile(keyFile, pem.EncodeToMemory(pemBlock), os.FileMode(0600))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	keyBytes, err := ioutil.ReadFile(keyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var block *pem.Block
+	var sigType x509.SignatureAlgorithm
+
+	block, _ = pem.Decode(keyBytes)
+
+	switch block.Type {
+	case "EC PRIVATE KEY":
+		privateKey, err = x509.ParseECPrivateKey(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+		ecdsaKey := privateKey.(*ecdsa.PrivateKey)
+		switch ecdsaKey.Curve.Params().BitSize {
+		case 521:
+			sigType = x509.ECDSAWithSHA512
+		case 384:
+			sigType = x509.ECDSAWithSHA384
+		default:
+			sigType = x509.ECDSAWithSHA256
+		}
+	case "RSA PRIVATE KEY":
+		privateKey, err = x509.ParsePKCS1PrivateKey(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+		rsaKey := privateKey.(*rsa.PrivateKey)
+		keySize := rsaKey.N.BitLen()
+		switch {
+		case keySize >= 4096:
+			sigType = x509.SHA512WithRSA
+		case keySize >= 3072:
+			sigType = x509.SHA384WithRSA
+		default:
+			sigType = x509.SHA256WithRSA
+		}
+	default:
+		return nil, fmt.Errorf("unsupported key type: %s", block.Type)
+	}
+
+	template := &x509.CertificateRequest{
+		Subject:            *subject,
+		SignatureAlgorithm: sigType,
+		DNSNames:           dnsSANs,
+		IPAddresses:        ipSANs,
```

Comment:
I would expect to see the "X509v3 Extended Key Usage" extension included here to request a cert valid for  "TLS Web Server Authentication" (RFC 5280, 4.2.1.12.  Extended Key Usage)